ampd/src/solana/verifier.rs

```diff
@@ -0,0 +1,234 @@
+use axelar_wasm_std::voting::Vote;
+use base64::Engine as _;
+use borsh::{BorshDeserialize, BorshSerialize};
+
+use base64::{self, engine::general_purpose};
+use tracing::warn;
+
+use crate::handlers::solana_verify_msg::Message;
+
+use super::{json_rpc::EncodedConfirmedTransactionWithStatusMeta, pub_key_wrapper::PubkeyWrapper};
+
+// Gateway program logs.
+// Logged when the Gateway receives an outbound message.
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone)]
+#[repr(u8)]
+enum GatewayEvent {
+    CallContract {
+        // Message sender.
+        sender: PubkeyWrapper,
+        destination_chain: Vec<u8>,
+        destination_address: Vec<u8>,
+        payload: Vec<u8>,
+        payload_hash: [u8; 32],
+    },
+}
+
+impl GatewayEvent {
+    // Try to parse a [`CallContractEvent`] out of a Solana program log line.
+    fn parse_log(log: &String) -> Option<Self> {
+        let cleaned_input = log
+            .trim()
+            .trim_start_matches("Program data:")
+            .split_whitespace()
+            .flat_map(decode_base64)
+            .next()?;
+        borsh::from_slice(&cleaned_input).ok()
+    }
+}
+
+impl PartialEq<&Message> for GatewayEvent {
+    fn eq(&self, msg: &&Message) -> bool {
+        match self {
+            GatewayEvent::CallContract {
+                sender: _,
+                destination_chain,
+                destination_address,
+                payload: _,
+                payload_hash,
+            } => {
+                let event_dest_addr = String::from_utf8(destination_address.to_owned());
+                let event_dest_chain = String::from_utf8(destination_chain.to_owned());
+
+                event_dest_addr.is_ok()
+                    && event_dest_chain.is_ok()
+                    && event_dest_addr.unwrap() == msg.destination_address
+                    && event_dest_chain.unwrap() == msg.destination_chain.to_string()
+                    && payload_hash.to_owned() == msg.payload_hash
+            }
+        }
+    }
+}
+
+#[inline]
+fn decode_base64(input: &str) -> Option<Vec<u8>> {
+    general_purpose::STANDARD.decode(input).ok()
+}
+
+pub fn verify_message(
+    source_gateway_address: &String, // TODO: check if sender is source_gateway_address
+    tx: &EncodedConfirmedTransactionWithStatusMeta,
+    message: &Message,
+) -> Vote {
+    if let None = tx.meta.log_messages {
+        // TODO: Log error/warn, because this event should contain log msgs?
+        return Vote::NotFound;
+    }
+
+    let program_data = tx.meta.log_messages.as_ref().unwrap();
+    let gw_event_parsed: Option<GatewayEvent> = program_data
+        .into_iter()
+        // TODO: Will find_map work with multiple msgs in transaction?
```

Comment:
I think `find_map` will stop on the first non `None` finding, ignoring next events. So actually we will be only working with one event. 

Under my eyes this is a only 1:1 verification logic. 1 axelar message (the one that enters the function as parameter) gets verified with one Solana event. I think we are constrained by the Axelar event handler architecture, as each event handler will react to only one event (message) in the axelar side. Taking into account this, i would probably determine in our side (Solana side) that only the first element of `program data` is the one that should contain the gateway event.

WDYT @ctoyan  ?

Reply:
The flow for msgs is as follows:
1. Some solana program sends a msg to our solana gateway
2. The solana gateway emits an event called CallContract
3. Relayer listens for it and catches it
4. Relayer constructs a msg (incl. solana tx signature) for the axelar protocol, based on the data in that caught event
5. It sends it for verification to ampd
6. Axelar contracts emit an event per msg
7. The code in this PR captures the event and compares the axelar msg against the solana blockchain and verifies that a tx with such signature (and data) exists.

AFAIK (not 100% sure), in step 1. the solana program can emit multiple msgs in a single tx (somehow.. i think in the logs).
The relayer should (doesn't currently) iterate over those multiple logs/whatever and creates multiple axelar msgs in step 4.
Then in step 7. when capturing each msg, ampd has to verify it against a single tx (all axelar messages will have the same tx signature in their ID, just a different index).
This means that in step 7. there is a possiblity that the solana tx logs contain multiple messages and the axelar message we're trying to verify, might be one of them. So there should be some iteration logic.

Does that make sense?

Reply:
owww, ok , i got it now. Many thanks for the explanation. 

> So there should be some iteration logic.

Yes, that at least check the incoming message from axelar is present in one of the `program logs` from the Solana tx. Let's iterate over them and do the msg comparison for each position. If we dont get at least one match, we will vote as failed such message.

Reply:
Let's just wait for Tiago to respond to [my questions in slack](https://eig3r.slack.com/archives/C05TRKNJDQS/p1708087004855879?thread_ts=1708074304.765779&cid=C05TRKNJDQS), to be sure how multiple messages can be represented in Solana, so we know what and how to iterate.

Reply:
@ctoyan  added the iteration logic at 0d484aa64ae775678270c87c13ff569a242ffa87

Reply:
Closing this as resolved. The new iteration logic is merged here now. Feel free to re-open.